src/txmempool.h

```diff
@@ -736,6 +739,55 @@ class CTxMemPool
      *  removal.
      */
     void removeUnchecked(txiter entry, MemPoolRemovalReason reason) EXCLUSIVE_LOCKS_REQUIRED(cs);
+public:
+    /** EpochGuard: RAII-style guard for using epoch-based graph traversal algorithms.
+     *     When walking ancestors or descendants, we generally want to avoid
+     * visiting the same transactions twice. Some traversal algorithms use
+     * std::set (or setEntries) to deduplicate the transaction we visit.
+     * However, use of std::set is algorithmically undesirable because it both
+     * adds an asymptotic factor of O(log n) to traverals cost and triggers O(n)
+     * more dynamic memory allocations.
+     *     In many algorithms we can replace std::set with an internal mempool
+     * counter to track the time (or, "epoch") that we began a traversal, and
+     * check + update a per-transaction epoch for each transaction we look at to
+     * determine if that transaction has not yet been visited during the current
+     * traversal's epoch.
+     *     Algorithms using std::set can be replaced on a one by one basis.
+     * Both techniques are not fundamentally incomaptible across the codebase.
+     * Generally speaking, however, the remaining use of std::set for mempool
+     * traversal should be viewed as a TODO for replacement with an epoch based
+     * traversal, rather than a preference for std::set over epochs in that
+     * algorithm.
+     */
+    class EpochGuard {
+        const CTxMemPool& pool;
```

Comment:
Hmm... I think that you're right that this code is somehow better, but TBH I don't really understand how the clang lock analyzing stuff works & don't know if I could personally maintain this code... Is there anything better to look at than http://releases.llvm.org/3.5.0/tools/clang/docs/ThreadSafetyAnalysis.html on how to understand this clang extension?

Any objection to this being a follow on PR?

I also have concern that this isn't compatible with future changes that may recursively use the epoch guard (there are algorithms that are recursive reuse safe, if you want to express a case where A has an epoch which calls B, and B creates a new Epoch which may re-traverse all of the mempool entries but A will not re-traverse anything traversed in B). You could imagine a use for this if I wanted to walk the whole mempool and traverse every connected component once, but at the same time.

This requires some modified behavior (e.g., making epoch guards aware of the epoch they are guarding), but can still work.

```c++
const auto epoch = GetFreshEpoch();
for (auto& elt : mapTx) {
    if (epoch.visited(elt)) continue;
    const auto epoch = epoch.SubEpoch();
    vector<txiter> component;
    GetAllAncestorsEpochAlreadyHeld(elt, component);
    GetAllDescendentsEpochAlreadyHeld(elt, component);
    DoSomethingWithComponent(component);
}
```

Reply:
If you're doing an "already held" thing, as far as clang goes that's just marking up the function with "EXCLUSIVE_LOCKS_REQUIRED(epoch)". I think something like:

```c++
class SubEpoch { uint64_t subepoch; }
class TxMemPool {
  ...
  GetAllAncestorsEpochAlreadyHeld(elt, vector<txiter>& component, const SubEpoch& subepoch) EXCLUSIVE_LOCKS_REQUIRED(epoch);
};
```

would make things pretty typesafe.

In the example you give, if you had a tx graph: 

 * A outputs A.0, A.1
 * B spends A.0, output B.0
 * C spends A.1, output C.0
 * D spends B.0, C.0

then if you traverse mapTx ordered as B,A,C,D you'll do {B,A,D} first, then {C,A,D} second; while if you'd traversed A or D first, you'd do {A,B,C,D} or {D,A,B,C} all in one go. I can't see why you'd end up with an algorithm like that.

Yeah, doing the refactor as a separate PR sounds fine

Reply:
@ajtowns so I think you're missing the algorithm that I suggested slightly, you'd actually in this case traverse all of {A, B, C, D} in the first sub-epoch, because you would be traversing all parents and children recursively.

So if you had another transaction E, with only confirmed inputs, and F which spends E.0, then you would output:

{{A, B, C, D}, {E, F}}

Hence being able to traverse the mempool and pick out connected components.

However, on further thought it occurs to me that this sort of algorithm would not require sub-epochs at all.

Reply:
Depends how you traversed mapTx -- if you picked `elt=B` first, then "C" would not be a parent or child of "B" so you'd only do {A,B,D} in the first subepoch. You'd need to be ordered by feerate or something weird for this to happen, though. But yeah, I can't think of something where sub-epochs is actually useful either. But whatever, `>=` vs `==` doesn't make much difference.

Reply:
follow-on PR is #18017

Reply:
Ah I think here's the miscommunication of what the algorithm "GetComponents" does. It's something like:

```
vector<vector<txiter>> components;
for (auto elt: mapTx) {
   vector<txiter> to_process;
   vector<txiter> processed;
   if (visited(elt)) continue;
   to_process.push_back(elt);
   while (to_process.size()) {
    auto todo = to_process.back();
    to_process.pop_back();
    processed.push_back(todo);
   for(auto y : GetMemPoolChildren(todo)) {
        if (visited(y)) continue;
        to_process.push_back(y);
   }
   for(auto x : GetMemPoolParents(todo)) {
       if (visited(x)) continue;
        to_process.push_back(x);
    }

}
    component.push_back(processed);
}
```

This lets you find all connected components I think, and passes the test of B-first. A is a parent of B, so then A goes into the to_process stack, then C is a child of A, so C goes onto the to_process stack.

Without epochs this would be much messier, but with epochs we can see that things only go onto the stack one time ever because of the visited guards.